### ISSUE TITLE
Spectrogram processing

### ISSUE THREAD

**Issue by RobinSchmidt** [RobinSchmidt/RS-MET#288]
i open a new thread to continue the discussion which started here:

https://github.com/RobinSchmidt/RS-MET/issues/280#issuecomment-525162273

but doesn't really belong to the main topic of this thread. ...so there's a class rsSpectrogram for converting an array of audio-samples into a spectrogram, represented as matrix of complex numbers and/or (re)synthesize an audio signal from such a spectrogram. in between analysis and resynthesis, one can apply arbitrary transformations to the spectrogram data. one of the simplemost things to do is filtering by zeroing out the bins above or below a certain cutoff point. the function `spectrogramFilter` in the TestsRosicAndRapt project demonstrates how this can be done.

...by writing this test function, i discovered a flaw in the underlying matrix class which unfortunately requires some inconvenient additional copying workaround (there are comments about this in the code). i think, i will soon replace this matrix class - which i wanted to do since some time anyway. i have now other ideas, how a proper implementation of a matrix class should look like...

**Comment by elanhickler** [comment]
Thanks for doing this.

Can you explain how blocksize, hopsize, trafosize would effect the filter?

**Comment by RobinSchmidt** [comment]
for filtering, the trafoSize should probably stay fixed at being equal to the blockSize. i think using zero-padding (i.e. trafoSize > blockSize) makes sense only for higher resolution analysis and visualization. it's actually a faux increase in frequency resulution - the spectrum is still washed out, you just get more spectral samples of it. hopSize = blockSize/2 is probably also best left as is. you can get similar faux time-resolution increase for visualization by choosing smaller (power-of-two) fractions here. the only parameter that really matters is the blockSize - it dials in a trade-off between time-resolution and frequency-resolution of the filter. it basically means that longer block-sizes lead to longer ringing filters that give better frequency separation

**Comment by elanhickler** [comment]
You have lowpass and highpass here. Wouldn't it be just as good to only do one and then subtract the result from the original?

**Comment by RobinSchmidt** [comment]
yes, that should work, too. i'm currently just not sure, if the subtraction will work correctly on the spectrogram level due to the flaws and/or bugs in the matrix class. but subtracting time-domain signals, should work

**Comment by elanhickler** [comment]
ok so something I kinda get but also totally don't get.

You have two classes, the spectrogram and the sinusoidal model stuff.

Let's say I want to create a simple denoiser. 0 out bins that are not above an amplitude threshold. Let's say I want to improve that denoiser by detecting harmonics. Let's say after x seconds the harmonics of the signal start dropping below the noise floor. This is where noise reduction start to become useless. What I really want is to do a guestimate of how long the harmonics last and just resynthesize from scratch. That's where the sinusoidal model comes in? Is it confusing to combine these two classes into a function? How's this going to work, etc. etc. This is just brainstorming.

**Comment by elanhickler** [comment]
quick question: is it ok to use float types with rsSpectrogram? Because the audio is in floats not doubles.

**Comment by RobinSchmidt** [comment]
> is it ok to use float types with rsSpectrogram? 

i didn't try that but don't see, why it should be a problem. you should just need an appropriate template instantiation

**Comment by elanhickler** [comment]
What's the demodulation thing about? What is modulation in this case?

**Comment by elanhickler** [comment]
![image](https://user-images.githubusercontent.com/9116575/64472815-ed2bc600-d118-11e9-81b4-b61d13838d41.png)

I get some distortion at the end of the file using the lowpass/highpass stuff. I resized my audio samples so that it has blocksize number of extra 0s at the end. That seemed to fix it.

**Comment by RobinSchmidt** [comment]
> What's the demodulation thing about? What is modulation in this case?

before analyzing a frame, an analysis window is applied and after (re)synthesizing a frame, a synthesis window is applied. depending on the choice of these windows (and the ration of blocksize and hopsize), there may be an amplitude modulation in the resynthesized signal - but this modulation is predictable and can be compensated for. this is the demodulation step.

in most implementations of spectrogram processing systems, the window-functions and blocksize-to-hopsize ratio is tuned such that this modulations does not occur (this happens when the (overlapped) products of analysis and synthesis windows sum up to unity). but i wanted to have more freedom in my choices of window-functions, so i  incorporated this demodulation step. this will do nothing (i.e. divide by one), in case of a choice where this overlap-to-unity condition is satisfied.  in other cases, it divides by the sum of the overlapped window-products

**Comment by RobinSchmidt** [comment]
> I get some distortion at the end of the file

:-O what is this?! i've never seen that. what are your settings? maybe i should try it with your sample?

**Comment by RobinSchmidt** [comment]
oh - i just noticed that the hopsize should be blockSize/4 and not blockSize/2 for the overlapped windows to sum up to a constant (with the Hann window). blockSize/2 would be appropriate only if the window would be applied once - but it's applied twice (in analysis and synthesis) - but with halving the hopsize, we again get a sum-up-to-constant property. the hann window is really nice in this respect

**Comment by elanhickler** [comment]
wtf, when I combine two files using JUCE classes the result is that there's errors around -135db. If I combine using REAPER, it is -inf, so, perfect. WHYYYYYY

**Comment by RobinSchmidt** [comment]
> It's an intermittent problem. Sometimes happens, sometimes doesn't

could it be related to the length of the buffer/file? if it's a *nice* number (maybe that an integer number of blocks fits in or something), it doesn't happen and other wise it does? or the other way around?

is the juce sample buffer single precision  and the reaper double or something? anyway, this is not related to my code, right?

**Comment by RobinSchmidt** [comment]
btw - i actually would - at the moment - recommend to prepend and append a blocksize worth of zeros before analysis and cutting it off after resynthesis (half of that should actually suffice but better be safe). because my block overlapping may produce fade-in/out effects at start and end (which you won't ever see because of the demodulation - they'll be compensated too - but it's probably better not having to compensate for anything)

...the spectrogram stuff is still very much under development

**Comment by RobinSchmidt** [comment]
ah - by the way -  in the function plotOverlappingWindowSum, i have made a plot of the overlapping windows and their sum, using hopSize=blockSize/4 with the Hann window. as you see, they sum up to unity:

![image](https://user-images.githubusercontent.com/1833099/64481189-36424f80-d1d7-11e9-9249-34ef351b56d1.png)

except for the fades at the ends. when you use the same overlap factor with a blackman window, you can clearly see the amplitude modulation:

![image](https://user-images.githubusercontent.com/1833099/64481217-eca63480-d1d7-11e9-86fd-5d1235ea62f4.png)

**Comment by elanhickler** [comment]
Yeah I'll have to experiment with double precision. I wonder if maybe using 64-bit wav files will force JUCE to use higher precision... or at least I should try 32 bit just to see what happens.

**Comment by RobinSchmidt** [comment]
>  This is basically the function of the sample tail extender but as you said, it doesn't use sinusoidal models? It only uses spectrogram stuff and bin amplitude? So it's not as good?

in some sense, it creates a (restricted version of a) sinusoidal model from the spectrogram data. for tail extension, i actually think, the general approach is quite appropriate - if it would be extended to estimating the decay rates from the data. but with the current one-decay-rate-for-all implementation, the tail sounds rather static and artificial.

**Comment by elanhickler** [comment]
Yep, using a 32-bit wav file solved the slight error.

OK, so how do you create a smooth spectral filter rather than brick wall? Do you simply reduce amplitude a little more for each frequency? Wouldn't you be able to hear the steppyness of the frequency rows?

**Comment by RobinSchmidt** [comment]
> how do you create a smooth spectral filter rather than brick wall? Do you simply reduce amplitude a little more for each frequency?

yes - i would probably linearly fade the amplitudes down over a certain number of bins. and/or maybe with a smoother (sin/cos) shaped function.  we would just have to take care, that complementary low- and highpass add up to unity. ...unless you get your highpass by subtraction - then, it would be prefect reconstruction, regardless. however, the highpass may not be an exact mirror image of the lowpass, if you use just any fade-function

> Wouldn't you be able to hear the steppyness of the frequency rows? Especially if the filter frequency was changing over time?

i guess, the overlap would smooth these steps out. what exactly do you mean to happen without time-modulation?

**Comment by RobinSchmidt** [comment]
ahh - i guess, i see what you mean: the rounding of the cutoff-bin to integer values. yeah...i could probably allow float numbers for that by scaling the last bin by a number between 0 and 1. i must think about that

**Comment by elanhickler** [comment]
btw, I don't see me needing a gradual filter. But I guess interpolating values is something that will be needed a lot, especially for a denoiser.

**Comment by elanhickler** [comment]
One thing a lot of my clients ask for is a matching filter, to for example make a loud violin sample sound like a softer violin sample, usually to correct mistakes in the recording process, or even to generate new samples that are in between recorded dynamics.

Do you know how matching stuff works?

**Comment by RobinSchmidt** [comment]
> Do you know how matching stuff works?

hmm - do you have some example product to show me exactly what you mean? if i'T about applying the spectral envelope of one signal to another signal, then i have done such things in the context of my master thesis (it was partially about spectral envelopes - and i implemented a vocoder based on this algorithm https://hal.archives-ouvertes.fr/hal-01161334/document)

**Comment by elanhickler** [comment]
Oh, well do YOU have some examples?

**Comment by RobinSchmidt** [comment]
ok - yeah - this is fun. this is me reading and then vocoding the title of my thesis (it's in german - "representation and modification of spectral envelopes of natural sounds based on formant models"):

www.rs-met.com/sounds/effects/ThesisVocoderCarrier.wav
www.rs-met.com/sounds/effects/ThesisVocoderModulator.wav
www.rs-met.com/sounds/effects/ThesisVocoderOutput.wav

**Comment by elanhickler** [comment]
So you can use this to convolve a soft guitar pluck with a loud guitar pluck and create something in between? Obviously won't sounds as good as the real thing, but just enough to be usable in music composition.

**Comment by RobinSchmidt** [comment]
well, it's vocoder - so the process is asymmetrical, so it's not like a morph (which i would expect to act somewhat symmetrical - and adjsutable) - carrier and modulator play different roles - but yeah, you get something "in between".

**Comment by elanhickler** [comment]
![image](https://user-images.githubusercontent.com/9116575/64482726-39741600-d1ac-11e9-84ef-0506bfea73ef.png)

http://www.elanhickler.com/transfer/ThesisVocoderOutput_VocodexPlugin.wav

This is Vocodex example but there's some problems  with it. It's great for a musical sound but your version sounds like a better starting point to improve on legibility and musicality.

**Comment by RobinSchmidt** [comment]
it has a totally different character. do you know how vocodex works? is it also spectrogram/stft based - or does it use the classical filter bank approach?

**Comment by elanhickler** [comment]
Vocodex is the best vocoder for music. Here's more exaggerated example
http://www.elanhickler.com/transfer/ThesisVocoderOutput_VocodexPlugin2.wav

So can you make a plugin with your vocoder?

**Comment by RobinSchmidt** [comment]
> So can you make a plugin with your vocoder?

the algorithm is implemented as a non-realtime matlab file. in principle, it could be turned into a realtime algorithm (in fact, in rosic, i already have some sort of framework for realtime spectral processors - that factors out all of the messy re-buffering business). but: this "true-spectral-envelope" algorithm is expensive. for a single frame, it iterates multiple fft/ifft roundtrips until convergence. not really good for realtime performance. however - i have some other ideas for simpler spectral envelope estimation algorithms - based on connecting peaks by lines or splines - which should probably give similar quality

**Comment by elanhickler** [comment]
Nevermind that! Explain how I can do some offline tests to see if it's at all viable for morphing two similar sound sources. I could make a function in SALT and use the scripting engine to play with it.

**Comment by RobinSchmidt** [comment]
ok - i just added all the .m files that i wrote back then for my thesis to my research repo:

https://github.com/RobinSchmidt/RS-MET-Research/tree/master/Prototypes/Octave/Thesis

to run the vocoder, you need to install octave:

https://www.gnu.org/software/octave/

and run this script:

https://github.com/RobinSchmidt/RS-MET-Research/blob/master/Prototypes/Octave/Thesis/trueEnvelopeVocoder.m

you should get exactly the output i posted above (the wavefile is dropped into the signals folder). put your input files there, too and modify the "audioread" calls appropriately.  note that you must also give it the fundamental frequencies fo carrier and modulator (well -  rough ballpark value is enough - it just scales the amount of envelope smoothing, if i remember correctly - mind you, i've not touched this code for 13 years!)

**Comment by elanhickler** [comment]
not sure what to do with this
![image](https://user-images.githubusercontent.com/9116575/64494989-d5ebf600-d248-11e9-91ba-69dc46ca1cdb.png)

**Comment by RobinSchmidt** [comment]
what? why is there html code in it. this is a matlab file!. first thing i'd recommend to do is to undock command-window from the center dock (i moved it to the right of the screen), so you can see the command window and editor window at once. then - why is your working folder  D:/Desktop? isn't it supposed to be something:/RS-MET-Research/Prototypes/Octave/Thesis

my screen looks like this (after running the script:)

![image](https://user-images.githubusercontent.com/1833099/64501245-df18a980-d2c0-11e9-8c0a-8ac7479fcf67.png)

**Comment by elanhickler** [comment]
I used the vocoder to create some in-between samples of guitar dynamics. It seems to work, maybe a few improvements could possibly be made, seems worth pursuing.

**Comment by RobinSchmidt** [comment]
interesting, non-standard use of a vocoder! :-O

**Comment by elanhickler** [comment]
http://www.elanhickler.com/transfer/guitar_dynamic_morph_3_to_5.mp3

**Comment by elanhickler** [comment]
Don't you have frames and bins reversed? Time frames are vertical (columns) frequency bins are horizontal (rows).

![image](https://user-images.githubusercontent.com/9116575/65055839-bfb6f780-d924-11e9-99ad-f2a354415bdf.png)

**Comment by elanhickler** [comment]
for some reason I am getting an infinte hang after deletion of the 2nd matrix in the code, I think it's the 2nd one if things are deleted in the order they appeared. The debugger isn't being helpful!

![image](https://user-images.githubusercontent.com/9116575/65060828-7919cb00-d92d-11e9-96bc-3b5ab519f82e.png)

I'm trying to zero out harmonics (small ranges of frequencies)

```cpp
for (int ch = 0; ch < channels; ++ch)
{
    // compute the complex spectrogram:
    WindowType W = WindowType::hanningZN;
    Spectrogram sp;
    sp.setBlockAndTrafoSize(blockSize, trafoSize);
    sp.setHopSize(hopSize);
    sp.setAnalysisWindowType(W);
    sp.setSynthesisWindowType(W);
    sp.setOutputDemodulation(true);
    Matrix s = sp.complexSpectrogram(origAudio->audio->getReadPointer(ch), samples);

    // workaround to create the deep copies
    int numFrames = s.getNumRows();
    int numBins = sp.getNumNonRedundantBins(); // == s.getNumColumns()        

    Matrix sl(numFrames, numBins);
    sl.copyDataFrom(s);

    vector<int> binsToZero;
    for (double cf = f; cf < sampleRate * 0.5;)
    {
        int lo = sp.frequencyToBinIndex(cf - fRange * 0.5, sampleRate);
        int hi = sp.frequencyToBinIndex(cf + fRange * 0.5, sampleRate);

        for (int b = lo; b < hi; ++b)
            binsToZero.push_back(b);

        cf += f;
    }

    // zero out harmonic bandwiths to get only noise
    for (int i = 0; i < numFrames; i++)
        for (int b : binsToZero)
            s(i, b) = 0;

    // subtract noise only from orignal to get harmonics only
    vec x = sp.synthesize(sl);
    auto ptr = harmAudio->audio->getWritePointer(ch);
    for (int s = 0; s < samples; ++s)
        ptr[s] -= x[s];

    // transfer noise only
    ptr = noisAudio->audio->getWritePointer(ch);
    for (int s = 0; s < samples; ++s)
        ptr[s] = x[s];
}
```

**Comment by elanhickler** [comment]
I just commented out the deletion function haha.  Look at my noise/harmonic spectrograms:
![ezgif-5-90b779e121d0 1](https://user-images.githubusercontent.com/9116575/65064016-5b9c2f80-d934-11e9-88bb-c79f1d0a57d2.gif)

I'm using a frequency bandwith of 15hz. That is INSANELY PRECISE! Your bidirectional filters could never do this.

Ok, I need to do another test though. It might be better to capture a large portion of frequencies rather than the smallest possible portion.

**Comment by elanhickler** [comment]
Here are the noise only waveforms:

tiny portion per harmonic:
![iZotope_RX_3_(64-bit)_2019-09-17_10-36-02](https://user-images.githubusercontent.com/9116575/65065296-01e93480-d937-11e9-93b1-211ccb87e1e3.png)

large portion per harmonic:
![iZotope_RX_3_(64-bit)_2019-09-17_10-33-06](https://user-images.githubusercontent.com/9116575/65065298-0281cb00-d937-11e9-926f-b72881740a9d.png)

So, the large portion is bad because you can see that the noise has a regularity to it, you can spot some oscillations. That's bad because it's going to interfere with the resynthesized harmonics when combining together. Looks like it's best to capture as tiny a range of frequencies per harmonic as possible.

**Comment by RobinSchmidt** [comment]
with "portion-per-harmonic" you mean the bandwidth of the bandpass filters to isolate the harmonics?

> the large portion is bad because you can see that the noise has a regularity to it, you can spot some oscillations

that makes sense. if the harmonic bandwidth gets wider, the "in-between" bandwidths of the noise bands get narrower. and narrow-bands have high sinusoidality/regularity

**Comment by RobinSchmidt** [comment]
> I think this is what I originally wanted to do from the beginning but never could due to not having flat enough filters.

hmmmm...actually, butterworth filters are quite nicely flat. ...but maybe not steep enough? ...in this case, one could go for elliptics at the cost of ripples (in passband and stopband). the FFT based filters have ripples, too. 

> That is INSANELY PRECISE! Your bidirectional filters could never do this.

so - this is good, right. i still cannot see, what fixed FFT filters can do that regular time-domain filters can't. maybe i should take the challenge of separating some signal with filters vs FFT. like a sort of battle experiment...haha

**Comment by elanhickler** [comment]
bidirectional filters can't change frequency right? But you could kinda move the filter around in a spectral process by changing bin volumes. Moving filters are needed.

**Comment by RobinSchmidt** [comment]
yes - it's difficult to make time-varying bidirectional IIR filters while preserving their desirable zero-phase properties. with FIR filters it would be easier but computationally expensive, such that you'd probably end up with doing some FFT based process here also (FFT convolution). but at them moment, we are talking about fixed filters, right? i mean the stuff you did above where you said that bidirectional filters could not do it. for time varying stuff - especially tracking harmonics frequencies -  i'd also opt for spectrogram based algorithms.

**Comment by elanhickler** [comment]
I'm having some issues with phaselocking, basically the same issues I had last time except I think things are sounding better and easier to use. I don't undertand how SampleModelling gets perfectly phaselocked samples.

http://www.elanhickler.com/transfer/horn_samples_phaselocked.rar

Included is the original: 
![image](https://user-images.githubusercontent.com/9116575/65083639-ee9d8f80-d95d-11e9-8355-0414ff1db514.png)


phaselocked example attempts to capture the phase per harmonic with 
```cpp
auto p = RAPT::rsSinePhaseAt<float>(x.data(), x.size(), x.size() * 0.5);
RAPT::rsRecreateSine<float>(x.data(), xNew.data(), x.size(), currentf, currentf, sampleRate, p, 0);
```
![image](https://user-images.githubusercontent.com/9116575/65083691-0ffe7b80-d95e-11e9-8a69-9707123cbc64.png)

phaselocked_singlePhase example sounds better because I don't take phase measurements but set [p]hase to "0" for all harmonics, whatever that means.

![image](https://user-images.githubusercontent.com/9116575/65083595-d0379400-d95d-11e9-850d-0319db8c6fd3.png)

I think `RAPT::rsSinePhaseAt` is not working well enough.

**Comment by elanhickler** [comment]
ALRIGHT YES! Taking the phase measurement precisely at the most energetic spot spectrally has improved the result:
![image](https://user-images.githubusercontent.com/9116575/65084679-ccf1d780-d960-11e9-9d98-5d5eed4a4007.png)

Now there's a weird issue with some amplitude modulation circled in red. No idea what that could be from.
![](https://cdn.discordapp.com/attachments/553292175534850049/623647959799693312/unknown.png)

**Comment by elanhickler** [comment]
Can you tell me how to, instead of zeroing out bins, manipulate the amplitude? First, retrieve the amplitude, and then change it to something else.

**Comment by RobinSchmidt** [comment]
each matrix entry is a complex number, so to get the amplitude, you can just call `std::abs` on it - the implementation for std::complex will extract the complex magnitude. you can also multiply the complex values by real numbers to change their magnitude. you may also want to look into `std::arg` and `std::polar`, if you want to deal with phase seaparately

**Comment by elanhickler** [comment]
What's an easy way to exponentially reduce the volume of a bin to 0? What equation should I use?

Also, should I use a sin function to feather the frequency vs bin volume? So you choose a center frequency, then above and below the volume is reduced a bit to 0 as well, so the filter isn't so flat.

**Comment by elanhickler** [comment]
I tried `sc(i, b) = abs(sc(i, b)) * (1 - x);`

but it's not working to fade out the amplitude, the fade is supposed to happen from 1 second to 2 seconds
![image](https://user-images.githubusercontent.com/9116575/65379913-2457b680-dc85-11e9-834b-22417c1c5770.png)

**Comment by RobinSchmidt** [comment]
> I'd need to convert a time range to a bin index range.

i don't understand that. time and frequency are independent dimensions

**Comment by elanhickler** [comment]
I got it working. Maybe I should have said numFrames / seconds

**Comment by RobinSchmidt** [comment]
yes - numFrames/seconds makes much more sense

**Comment by elanhickler** [comment]
I'll explain what I'm up to. I'm fading it/removing all audio content except multiples of a frequency. That isolates harmonics. It drastically reduces noise and clicks. The problem is that you hear a buzzing sound because there's still noise in the upper frequencies mostly. To solve this I put a sweeping-down lowpass filter so the harmonics decay in a way that sounds natural. It's just guesswork at that point, by ear. But here's the kicker: you could not use a filter on a normal signal because it would actually sound like a seeping lowpass filter because of broadband noise. It would sound like "showww". Since everything but harmonics is isolates using a filter is viable and sounds correct.

**Comment by RobinSchmidt** [comment]
```cpp
    // workaround to create the deep copies
    Matrix sl(numFrames, numBins);
    sl.copyDataFrom(s);
```
meanwhile, i have implemented my new matrix class and i have updated my spectrogram stuff to use it. now you can simply write:
```cpp
    Matrix sl = s;
```
the new  class is temporarily named `rsMatrixNew`, so you may have to update the typedef that is not shown your code. eventually, i want to retire the old implementation and then the new class will be just named `rsMatrix` again. but during transition (while the old class is still in use), i need this temporary name for distinction

**Comment by elanhickler** [comment]
I don't know wtf is going on, but this is really bad. I can't extract harmonics at a crucial time. See code: https://github.com/RobinSchmidt/RS-MET/issues/288#issuecomment-532298095

It's not working.

I think you need to demonstrate that your code is working to zero-out bandwidths of frequencies.

**Comment by elanhickler** [comment]
this isn't working, the resynthesis somehow is resulting in exactly the original audio despite zeroing out bins.

```cpp
vector<int> binsToZero;
			for (double cf = f; cf < sampleRate * 0.5;)
			{
				int lo = sp.frequencyToBinIndex(cf - fRange * 0.5, sampleRate);
				int hi = sp.frequencyToBinIndex(cf + fRange * 0.5, sampleRate);

				for (int b = lo; b < hi; ++b)
					binsToZero.push_back(b);

				cf += f;
			}

			// zero out harmonic bandwiths to get only noise
			for (int i = 0; i < numFrames; i++)
				for (int b : binsToZero)
					s(i, b) = 0;

			// subtract noise only from orignal to get harmonics only
			vec x = sp.synthesize(sl);
			auto ptr = harmAudio->audio->getWritePointer(ch);
			for (int s = 0; s < samples; ++s)
				ptr[s] -= x[s];

			// transfer noise only
			ptr = noiseAudio->audio->getWritePointer(ch);
			for (int s = 0; s < samples; ++s)
				ptr[s] = x[s];
```

**Comment by elanhickler** [comment]
I went back to September 20 commits, same result. What the hell am I doing wrong? I'm using tested code, it worked before. How am I zeroing out every bin of audio when my vector of bins is not every bin?

**Comment by RobinSchmidt** [comment]
>  I was resynthesizing the wrong fft object.......................... I don't understand this code too well

what is an fft object? do you mean the spectrogram matrix? if you think, the code is hard to use and have suggestions for the improvement of the API, let me know. in fact, i'm contemplating a general overhaul of the API - after watching this (very entertaining) video:

https://www.youtube.com/watch?v=5tg1ONG18H8

my code grew organically over almost two decades now and was not at all consistently "designed" in any way (aside from *loosely* following some conventions that i made up), so it definitely has its quirks

**Comment by elanhickler** [comment]
```cpp
Matrix s = sp.complexSpectrogram(origAudio->audio->getReadPointer(ch), samples);

// workaround to create the deep copies
int numFrames = s.getNumRows();
int numBins = sp.getNumNonRedundantBins(); // == s.getNumColumns()        

Matrix sl = s;
```

this is what messed me up, because I don't understand why we throw away the first matrix.

**Comment by RobinSchmidt** [comment]
oh - you mean, you want to work directly on the matrix `s` instead of on a copy `sl` as in "in-place" processing? sure, if you don't need the unprocessed matrix anymore for anything else, you can totally do that. the code from which you copy-and-pasted this, needs the original matrix later for plotting - but you may not

**Comment by elanhickler** [comment]
I don't get it. Why would you need an unprocessed matrix for plotting? Wouldn't you want to plot the processed version to see the results?

Here's my code again:

```cpp
const int channels = origAudio->getNumChannels();
const int sampleRate = origAudio->getSampleRate();
const int samples = origAudio->getNumFrames();
const int bitDepth = origAudio->getBitDepth();
const double maxf = std::min<double>(20000, sampleRate * 0.5);
const int nHarmonics = int(maxf / f);

for (int ch = 0; ch < channels; ++ch)
{
    // compute the complex spectrogram:
    WindowType W = WindowType::hanningZN;
    Spectrogram sp;
    sp.setBlockAndTrafoSize(blockSize, trafoSize);
    sp.setHopSize(hopSize);
    sp.setAnalysisWindowType(W);
    sp.setSynthesisWindowType(W);
    sp.setOutputDemodulation(true);
    Matrix s = sp.complexSpectrogram(origAudio->audio->getReadPointer(ch), samples);

    // workaround to create the deep copies
    int numFrames = s.getNumRows();
    int numBins = sp.getNumNonRedundantBins(); // == s.getNumColumns()        

    Matrix sl = s;

    vector<int> binsToZero;
    for (double cf = f; cf < sampleRate * 0.5;)
    {
        int lo = sp.frequencyToBinIndex(cf - fRange * 0.5, sampleRate);
        int hi = sp.frequencyToBinIndex(cf + fRange * 0.5, sampleRate);

        for (int b = lo; b < hi; ++b)
            binsToZero.push_back(b);

        cf += f;
    }

    // zero out harmonic bandwiths to get only noise
    for (int i = 0; i < numFrames; i++)
        for (int b : binsToZero)
            sl(i, b) = 0;

    // subtract noise only from orignal to get harmonics only
    vec x = sp.synthesize(sl);
    auto ptr = harmAudio->audio->getWritePointer(ch);
    for (int s = 0; s < samples; ++s)
        ptr[s] -= x[s];

    // transfer noise only
    ptr = noiseAudio->audio->getWritePointer(ch);
    for (int s = 0; s < samples; ++s)
        ptr[s] = x[s];
}

if (noiseOutPath.isNotEmpty())
    noiseAudio->renderFile(noiseOutPath);
if (harmOutPath.isNotEmpty())
    harmAudio->renderFile(harmOutPath);

return var::undefined;
```

**Comment by RobinSchmidt** [comment]
> Why would you need an unprocessed matrix for plotting? Wouldn't you want to plot the processed version to see the results? And yes, I don't need to plot results. Can I get rid of the 2nd matrix?

because in my test, i plot all three spectrograms - original, lowpassed and highpassed. sure, i could have interleaved the processing and plotting code to get rid of at least of the matrices - but this is not production code, so i'd rather focus on readability than efficiency.

>  Also, I want to move numFrames/numBins and vector<int> binsToZero outside of the loop because that only needs to be calculated once.

i don't see any obvious problems with that - numFrames/numBins seem to be used only once, so you may as well replace their usage directly with the function call. although i think, looping up to numFrames is more readable than to s.getNumRows(), though. matter of taste.  i wonder, why you bother with the binsToZero vector at all and not just loop directly `for(b = lo; b <= hi; b++)`

**Comment by RobinSchmidt** [comment]
i just renamed `rsSpectrogram` to `rsSpectrogramProcessor ` - and the member function `complexSpectrogram` to `getComplexSpectrogram`. i'm planning to introduce a new class `rsComplexSpectrogram` which should be a subclass of `rsMatrix<std::complex<T>>`. then you would have something like `rsSpectrogramProcessor::getComplexSpectrogram` that returns an object of class `rsComplexSpectrogram::complexSpectrogram` rather than `rsSpectrogram` returning `rsMatrix`. that makes much more sense, API-wise

**Comment by elanhickler** [comment]
good, push changes!

**Comment by elanhickler** [comment]
```cpp
for (int ch = 0; ch < channels; ++ch)
{
    // compute the complex spectrogram:
    WindowType W = WindowType::hanningZN;
    Spectrogram sp;
    sp.setBlockAndTrafoSize(blockSize, trafoSize);
    sp.setHopSize(hopSize);
    sp.setAnalysisWindowType(W);
    sp.setSynthesisWindowType(W);
    sp.setOutputDemodulation(true);
    Matrix mat = sp.complexSpectrogram(origAudio->audio->getReadPointer(ch), samples);

    // zero out harmonic bandwiths to get only noise
    for (int i = 0; i < mat.getNumRows(); i++)
    {
        for (double cf = f; cf < sampleRate * 0.5;)
        {
            int lo = sp.frequencyToBinIndex(cf - fRange * 0.5, sampleRate);
            int hi = sp.frequencyToBinIndex(cf + fRange * 0.5, sampleRate);

            for (int b = lo; b < hi; ++b)
                mat(i, b) = 0;

            cf += f;
        }                
    }

    // subtract noise only from orignal to get harmonics only
    vec x = sp.synthesize(mat);
    auto ptr = harmAudio->audio->getWritePointer(ch);
    for (int s = 0; s < samples; ++s)
        ptr[s] -= x[s];

    // transfer noise only
    ptr = noiseAudio->audio->getWritePointer(ch);
    for (int s = 0; s < samples; ++s)
        ptr[s] = x[s];
}
```

**Comment by RobinSchmidt** [comment]
i already pushed them. i would also drag out all the setup-work for sp out of the loop over the channels - execept for the last line `mat = ...` - because nothing really changes - you just do all the exact same setup-operations numChannels times when doing it in the loop

**Comment by elanhickler** [comment]
build error, i fixed it like this... not sure if it's correct.
![image](https://user-images.githubusercontent.com/9116575/69686265-7e596b80-107c-11ea-81ee-9db77bbaf1c1.png)

**Comment by RobinSchmidt** [comment]
oh - ok, yes - thanks. i fixed it. i didn't hit this on "build" - only now on "rebuild". apparently, some minimum-rebuild-cache is VS was messed up.
